src/backend/optimizer/path/allpaths.c

```diff
@@ -2125,15 +2125,36 @@ set_cte_pathlist(PlannerInfo *root, RelOptInfo *rel, RangeTblEntry *rte)
 			subplan = subquery_planner(cteroot->glob, subquery, cteroot, cte->cterecursive,
 									   tuple_fraction, &subroot, config);
 
-			cteplaninfo->shared_plan = prepare_plan_for_sharing(cteroot, subplan);
+			cteplaninfo->subplan = (subplan->flow->locustype != CdbLocusType_SegmentGeneral) ?
+			                       prepare_plan_for_sharing(cteroot, subplan) :
+			                       subplan;
+
 			cteplaninfo->subroot = subroot;
 		}
 
 		/*
 		 * Create another ShareInputScan to reference the already-created
-		 * subplan.
+		 * subplan if not avoiding sharing. Sharing SegmentGeneral subplan may
+		 * lead to deadlock when executed with 1-gang and joined with Hashed.
+		 * Forcing locus to Replicated leads to redundant motions if we join
+		 * shared plan with another SegmentGeneral node. Thus, we should avoid
+		 * sharing SegmentGeneral subplans.
```

Comment:
information about an alternative solution does not make sense in the comments to the code.